File: distributed-rate-limiter/src/rate-limiter/sliding_window_counter.ts
import { RateLimiter, slidingWindowCounterConfig, slidingWindowCounterResponse } from './typeModel';
import * as _ from '../constants';
import { redisClient } from '../redis'

/**
 * 滑動窗口視窗限流器
 * timeWindows: number // 時間窗口大小 (秒)
 * maxRequests: number // 最大請求數
 * startTime: number // 開始時間 (秒) --錨定時間點
 * prevCount: number // 前一個窗口的計數
 * currentCount: number // 當前窗口的計數
 * lastRequestWindow: number // 最後請求窗口
 */
export default class SlidingWindowCounter implements  RateLimiter {
    async run(key: string): Promise<slidingWindowCounterResponse> {
        return Boolean(await (redisClient as any)[_.SLIDING_WINDOW_COUNTER](1, key, ''));
    }

    async set(key: string, config: slidingWindowCounterConfig): Promise<boolean> {
        const startTime = Math.floor(Date.now() / 1000);
        return Boolean(await redisClient.hset(key,
            'timeWindows', config.timeWindows,
            'maxRequests', config.maxRequests,
            'startTime', startTime,
            'prevCount', 0,
            'currentCount', 0,
            'lastRequestWindow', 0
        ));
    }

    async delete(key: string): Promise<boolean> {
        return Boolean(await redisClient.del(key));
    }
}

File: distributed-rate-limiter/src/redis/lua/sliding_window_counter.lua
local key = KEYS[1]

local arr = redis.call('HGETALL', key)
-- 沒有限流，直接通過
if #arr == 0 then
    return 1
end

-- 將數組轉換為對象
local obj = {}
for i = 1, #arr, 2 do
    obj[arr[i]] = tonumber(arr[i + 1])
end

-- 轉為自定義時間窗口
local time = tonumber(redis.call('TIME')[1]) - obj.startTime
local current_window = math.floor(time / obj.timeWindows)
local ratio = 1 - (time % obj.timeWindows) / obj.timeWindows

-- 更新時間軸
if obj.lastRequestWindow + 1 == current_window then
    obj.prevCount = obj.currentCount
    obj.currentCount = 0
elseif obj.lastRequestWindow + 1 < current_window then
    obj.prevCount = 0
    obj.currentCount = 0
end

-- 加權請求數 = (上一個視窗的請求數 * 滑動比例) + 當前視窗的請求數
-- 滑動比例：1 - (當前時間戳佔當前視窗的比例)
local weighted = (obj.prevCount * ratio) + obj.currentCount
if weighted < obj.maxRequests then
    redis.call('HSET', key, 'lastRequestWindow', current_window, 'prevCount', obj.prevCount, 'currentCount', obj.currentCount + 1)
    return 1
end
return 0